code/Chapter5/test_dependency_injection.cpp

```diff
@@ -0,0 +1,84 @@
+#include <iostream>
+#include <cmath>
+#include "gtest/gtest.h"
+
+/// @brief A "complex" calculation of the norm.
+/// @param array The vector to calculate the norm for.
+/// @return The actual norm of the vector.
+double calculate_norm(std::vector<int> array)
+{
+  double output{0};
+  for (int i{0}; i < array.size(); ++i)
+  {
+    output = array[i] * array[i];
+  }
+  return std::sqrt(output);
+}
+
+/// @brief Just a stub, so we can get a simple value to test things with.
+/// @param array The vector to calculate the norm for.
+/// @return The norm, a dummy value in this case.
+double norm_stub(std::vector<int> array)
+{
+  return 10.0;
+}
```

Comment:
```suggestion
double norm_stub(std::vector<int> array)
{
  return //Actual v2 norm;
}
```

Since we are calculating L2 norm of the vector, we can use the actual/realistic value to return from stub. Although 10 might also serve the purpose but it is not a good idea to check code against incorrect value/s.

Reply:
I disagree. The whole point of the example is to show a simplified version of the norm, therefore returning a constant, known value (that would be a stub, in this case). Giving a proper implementation of the norm defeats the purpose of the example.

Reply:
I would highly disagree (with no intentions of criticizing offending ) with the value that has been chosen. This is the only reason I made my comment at the first place.

Yes, you are right that we should not be bothered about exact norm implementation for a stub as that would defeat the purpose. 

Having said that, it will also not make sense to use responses/values which are complete outliers or cannot happen. That might invalidate your tests also.

How about if someone wants to use -10 as value. Our tests will pass. However, negative values cannot be returned from a norm function. What would be your take on negative value usage in the stub?

Reply:
I wouldn't have any problem with it. 

I think you're mixing unit testing with functional testing. In the second case, having not only a meaningful stub, but an actual norm calculation will be essential. 

But in unit testing we just need to check that this function - and this function only, not anything called by it - its doing its job. This function is dividing a vector by the output of a function that takes the vector as input, and that's what needs testing. That the norm is a positive or negative number is irrelevant for the unit test of this function. Ensuring that the norm is positive will be the purpose of unit test written for the functions that calculate the norm, not for this one.